hail/src/main/scala/is/hail/expr/ir/Infer.scala

```diff
@@ -1,89 +1,101 @@
 package is.hail.expr.ir
 
-import is.hail.expr.types.{virtual, _}
+import is.hail.expr.types.physical._
 import is.hail.expr.types.virtual._
+import is.hail.utils._
 
 object Infer {
-  def apply(ir: InferIR): Type = {
+  def apply(ir: InferIR): PType = {
     ir match {
       case If(cond, cnsq, altr) =>
         assert(cond.typ.isOfType(TBoolean()))
         assert(cnsq.typ.isOfType(altr.typ))
-        if (cnsq.typ != altr.typ)
-          cnsq.typ.deepOptional()
+        if (cnsq.pType != altr.pType)
+          PType.canonical(cnsq.typ.deepOptional())
         else
-          cnsq.typ
+          cnsq.pType
       case Let(name, value, body) =>
-        body.typ
+        body.pType
       case ApplyBinaryPrimOp(op, l, r) =>
-        BinaryOp.getReturnType(op, l.typ, r.typ)
+        PType.canonical(BinaryOp.getReturnType(op, l.typ, r.typ)).setRequired(l.pType.required && r.pType.required)
       case ApplyUnaryPrimOp(op, v) =>
-        UnaryOp.getReturnType(op, v.typ)
+        PType.canonical(UnaryOp.getReturnType(op, v.typ)).setRequired(v.pType.required)
       case ApplyComparisonOp(op, l, r) =>
         assert(l.typ isOfType r.typ, s"${l.typ.parsableString()} vs ${r.typ.parsableString()}")
-        TBoolean()
+        PBoolean(l.pType.required && r.pType.required)
       case ArrayRef(a, i) =>
         assert(i.typ.isOfType(TInt32()))
-        -coerce[TArray](a.typ).elementType
+        coerce[PArray](a.pType).elementType.setRequired(a.pType.required && i.pType.required)
       case ArraySort(a, ascending, _) =>
         assert(ascending.typ.isOfType(TBoolean()))
-        a.typ
+        val et = coerce[PArray](a.pType).elementType
+        PArray(et, a.pType.required)
       case ToSet(a) =>
-        TSet(coerce[TArray](a.typ).elementType)
+        val et = coerce[PArray](a.pType).elementType
+        PSet(et, a.pType.required)
       case ToDict(a) =>
-        val elt = coerce[TBaseStruct](coerce[TArray](a.typ).elementType)
-        TDict(elt.types(0), elt.types(1))
+        val elt = coerce[PBaseStruct](coerce[PArray](a.pType).elementType)
+        PDict(elt.types(0), elt.types(1), a.pType.required)
       case ToArray(a) =>
-        TArray(coerce[TContainer](a.typ).elementType)
+        val et = coerce[PContainer](a.pType).elementType
+        PArray(et, a.pType.required)
       case GroupByKey(collection) =>
-        val elt = coerce[TBaseStruct](coerce[TArray](collection.typ).elementType)
-        TDict(elt.types(0), TArray(elt.types(1)))
+        val elt = coerce[PBaseStruct](coerce[PArray](collection.pType).elementType)
+        // FIXME requiredness
+        PDict(elt.types(0), PArray(elt.types(1), required = false), collection.pType.required)
       case ArrayMap(a, name, body) =>
-        TArray(-body.typ, a.typ.required)
+        // FIXME: requiredness artifact of former IR bug, remove when possible
+        PArray(body.pType.setRequired(false), a.typ.required)
```

Comment:
Same for this.  Former IR bug?  If it's former, why's this needed?  What was the bug?

Reply:
The "bug" was us not handling requiredness correctly in many contexts. I don't want to start generating requiredness until non-canonical ptypes are ready.